File: Framework Plugins/WADV.PhysicsModule/Point2.vb
''' <summary>
''' 表示一个二维点
''' </summary>
Public Structure Point2
    ''' <summary>
    ''' 点的X轴坐标
    ''' </summary>
    Public X As Double
    ''' <summary>
    ''' 点的Y轴坐标
    ''' </summary>
    Public Y As Double

    ''' <summary>
    ''' 获得一个新的二维点
    ''' </summary>
    ''' <param name="x">点的X轴坐标</param>
    ''' <param name="y">点的Y轴坐标</param>
    Public Sub New(x As Double, y As Double)
        Me.X = x
        Me.Y = y
    End Sub

    Public Shared Operator +(a As Point2, b As Point2) As Point2
        Return New Point2(a.X + b.X, a.Y + b.Y)
    End Operator

    Public Shared Operator +(a As Point2, b As Double) As Point2
        Return New Point2(a.X + b, a.Y + b)
    End Operator

    Public Shared Operator +(a As Point2, b As Integer) As Point2
        Return New Point2(a.X + b, a.Y + b)
    End Operator

    Public Shared Operator +(a As Point2, b As Vector2) As Point2
        Return New Point2(a.X + b.X, a.Y + b.Y)
    End Operator

    Public Shared Operator -(a As Point2, b As Point2) As Point2
        Return New Point2(a.X - b.X, a.Y - b.Y)
    End Operator

    Public Shared Operator -(a As Point2, b As Double) As Point2
        Return New Point2(a.X - b, a.Y - b)
    End Operator

    Public Shared Operator -(a As Point2, b As Integer) As Point2
        Return New Point2(a.X - b, a.Y - b)
    End Operator

    Public Shared Operator -(a As Point2, b As Vector2) As Point2
        Return New Point2(a.X - b.X, a.Y - b.Y)
    End Operator

    Public Shared Operator *(a As Point2, b As Point2) As Point2
        Return New Point2(a.X * b.X, a.Y * b.Y)
    End Operator

    Public Shared Operator *(a As Point2, b As Double) As Point2
        Return New Point2(a.X * b, a.Y * b)
    End Operator

    Public Shared Operator *(a As Point2, b As Integer) As Point2
        Return New Point2(a.X * b, a.Y * b)
    End Operator

    Public Shared Operator *(a As Point2, b As Vector2) As Point2
        Return New Point2(a.X * b.X, a.Y * b.Y)
    End Operator

    Public Shared Operator /(a As Point2, b As Point2) As Point2
        Return New Point2(a.X / b.X, a.Y / b.Y)
    End Operator

    Public Shared Operator /(a As Point2, b As Double) As Point2
        Return New Point2(a.X / b, a.Y / b)
    End Operator

    Public Shared Operator /(a As Point2, b As Integer) As Point2
        Return New Point2(a.X / b, a.Y / b)
    End Operator

    Public Shared Operator /(a As Point2, b As Vector2) As Point2
        Return New Point2(a.X / b.X, a.Y / b.Y)
    End Operator

    Public Shared Operator =(a As Point2, b As Point2) As Boolean
        Return a.X = b.X AndAlso a.Y = b.Y
    End Operator

    Public Shared Operator <>(a As Point2, b As Point2) As Boolean
        Return Not a = b
    End Operator

    Public Shared Operator >(a As Point2, b As Point2) As Boolean
        Return (a.X * a.X + a.Y * a.Y) > (b.X * b.X + b.Y * b.Y)
    End Operator

    Public Shared Operator <(a As Point2, b As Point2) As Boolean
        Return (a.X * a.X + a.Y * a.Y) < (b.X * b.X + b.Y * b.Y)
    End Operator

    Public Shared Operator >=(a As Point2, b As Point2) As Boolean
        Return (a.X * a.X + a.Y * a.Y) >= (b.X * b.X + b.Y * b.Y)
    End Operator

    Public Shared Operator <=(a As Point2, b As Point2) As Boolean
        Return (a.X * a.X + a.Y * a.Y) <= (b.X * b.X + b.Y * b.Y)
    End Operator

    Public Shared Widening Operator CType(v As Integer) As Point2
        Return New Point2(v, v)
    End Operator

    Public Shared Widening Operator CType(v As Double) As Point2
        Return New Point2(v, v)
    End Operator

    ''' <summary>
    ''' 确定两个二维点的X轴坐标是否相同
    ''' </summary>
    ''' <param name="a">第一个二维点</param>
    ''' <param name="b">第二个二维点</param>
    ''' <returns></returns>
    Public Shared Function EqualX(a As Point2, b As Point2) As Boolean
        Return a.X = b.X
    End Function

    ''' <summary>
    ''' 确定两个二维点的Y轴坐标是否相同
    ''' </summary>
    ''' <param name="a">第一个二维点</param>
    ''' <param name="b">第二个二维点</param>
    ''' <returns></returns>
    Public Shared Function EqualY(a As Point2, b As Point2) As Boolean
        Return a.Y = b.Y
    End Function

    ''' <summary>
    ''' 零点
    ''' </summary>
    ''' <returns></returns>
    Public Shared ReadOnly Property Zero As Point2 = New Point2 With {.X = 0.0, .Y = 0.0}
End Structure

File: WADV.Core.NETCore/Utility/Point2.cs
namespace WADV.Core.NETCore.Utility
{
    /// <summary>
    /// 2D point
    /// </summary>
    public struct Point2
    {
        /// <summary>
        /// X coordinate
        /// </summary>
        public double X;
        /// <summary>
        /// Y coordinate
        /// </summary>
        public double Y;

        /// <summary>
        /// Get a new 2D point
        /// </summary>
        /// <param name="x">X coordinate</param>
        /// <param name="y">Y coordinate</param>
        public Point2(double x, double y)
        {
            X = x;
            Y = y;
        }

        public static Point2 operator +(Point2 a, Point2 b)
        {
            return new Point2(a.X + b.X, a.Y + b.Y);
        }
        public static Point2 operator +(Point2 a, double b)
        {
            return new Point2(a.X + b, a.Y + b);
        }
        public static Point2 operator +(Point2 a, int b)
        {
            return new Point2(a.X + b, a.Y + b);
        }
        public static Point2 operator +(Point2 a, Vector2 b)
        {
            return new Point2(a.X + b.X, a.Y + b.Y);
        }
        public static Point2 operator -(Point2 a, Point2 b)
        {
            return new Point2(a.X - b.X, a.Y - b.Y);
        }
        public static Point2 operator -(Point2 a, double b)
        {
            return new Point2(a.X - b, a.Y - b);
        }
        public static Point2 operator -(Point2 a, int b)
        {
            return new Point2(a.X - b, a.Y - b);
        }
        public static Point2 operator -(Point2 a, Vector2 b)
        {
            return new Point2(a.X - b.X, a.Y - b.Y);
        }
        public static Point2 operator *(Point2 a, Point2 b)
        {
            return new Point2(a.X * b.X, a.Y * b.Y);
        }
        public static Point2 operator *(Point2 a, double b)
        {
            return new Point2(a.X * b, a.Y * b);
        }
        public static Point2 operator *(Point2 a, int b)
        {
            return new Point2(a.X * b, a.Y * b);
        }
        public static Point2 operator *(Point2 a, Vector2 b)
        {
            return new Point2(a.X * b.X, a.Y * b.Y);
        }
        public static Point2 operator /(Point2 a, Point2 b)
        {
            return new Point2(a.X / b.X, a.Y / b.Y);
        }
        public static Point2 operator /(Point2 a, double b)
        {
            return new Point2(a.X / b, a.Y / b);
        }
        public static Point2 operator /(Point2 a, int b)
        {
            return new Point2(a.X / b, a.Y / b);
        }
        public static Point2 operator /(Point2 a, Vector2 b)
        {
            return new Point2(a.X / b.X, a.Y / b.Y);
        }
        public static bool operator ==(Point2 a, Point2 b)
        {
            return a.X.Equals(b.X) && a.Y.Equals(b.Y);
        }
        public static bool operator !=(Point2 a, Point2 b)
        {
            return !(a == b);
        }
        public static bool operator >(Point2 a, Point2 b)
        {
            return a.X > b.X && a.Y > b.Y;
        }
        public static bool operator <(Point2 a, Point2 b)
        {
            return a.X > b.X && a.Y < b.Y;
        }
        public static bool operator >=(Point2 a, Point2 b)
        {
            return a.X > b.X && a.Y >= b.Y;
        }
        public static bool operator <=(Point2 a, Point2 b)
        {
            return a.X > b.X && a.Y <= b.Y;
        }
        public static implicit operator Point2(int v)
        {
            return new Point2(v, v);
        }
        public static implicit operator Point2(double v)
        {
            return new Point2(v, v);
        }

        /// <summary>
        /// Indicate another point's X coordinate is same or not with this point
        /// </summary>
        /// <param name="target">Target</param>
        /// <returns></returns>
        public bool EqualX(Point2 target)
        {
            return X.Equals(target.X);
        }
        /// <summary>
        /// Indicate another point's X coordinate is same or not with this point
        /// </summary>
        /// <param name="target">Target</param>
        /// <returns></returns>
        public bool EqualY(Point2 target)
        {
            return Y.Equals(target.Y);
        }

        /// <summary>
        /// Zero point
        /// </summary>
        /// <returns></returns>
        public static Point2 Zero { get; } = new Point2(0, 0);

        public bool Equals(Point2 other)
        {
            return X.Equals(other.X) && Y.Equals(other.Y);
        }
        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            return obj is Point2 && Equals((Point2)obj);
        }
        public override int GetHashCode()
        {
            unchecked
            {
                return (X.GetHashCode() * 397) ^ Y.GetHashCode();
            }
        }
    }
}